app/assets/javascripts/discourse/app/components/translation.gjs

```diff
@@ -0,0 +1,202 @@
+import Component from "@glimmer/component";
+import { action } from "@ember/object";
+import { getOwner } from "@ember/owner";
+import curryComponent from "ember-curry-component";
+import TranslationPlaceholder from "discourse/components/translation-placeholder";
+import uniqueId from "discourse/helpers/unique-id";
+import { isProduction } from "discourse/lib/environment";
+import I18n, { i18n, I18nMissingInterpolationArgument } from "discourse-i18n";
+
+/**
+ * Provides the ability to interpolate both strings and components into translatable strings.
+ * This component allows for complex i18n scenarios where you need to embed interactive
+ * components within translated text.
+ *
+ * If you don't require this functionality, use the standard i18n() function.
+ *
+ * @component Translation
+ *
+ * @template Usage example:
+ * ```gjs
+ * // Translation key: "some.translation.key" = "Welcome, %{username}! The date is %{shortdate}!"
+ * <Translation
+ *   @key="some.translation.key"
+ *   @options={{hash shortdate=shortDate}}
+ *   as |Placeholder|
+ * >
+ *   <Placeholder @name="username">
+ *     <UserLink @user={{user}}>{{user.username}}</UserLink>
+ *   </Placeholder>
+ * </Translation>
+ * ```
+ *
+ * @param {String} key - The i18n translation key to use
+ * @param {Object} [options] - Hash of options to pass to the i18n function for string interpolation
+ */
+export default class Translation extends Component {
+  /**
+   * Processes the translation string and returns all of the rendering data.
+   *
+   * @returns {Object}
+   */
+  get textAndPlaceholders() {
+    const optionsArg = this.args.options || {};
+
+    // Find all of the placeholders in the string we're looking at.
+    const message = I18n.findTranslationWithFallback(this.args.key, {
+      ...optionsArg,
+    });
+    const placeholderAppearance = I18n.findPlaceholders(message);
+
+    // We only need to keep the placeholders that aren't being handled by those passed in @options.
+    Object.keys(optionsArg).forEach((stringPlaceholder) =>
+      placeholderAppearance.delete(stringPlaceholder)
+    );
+
+    const placeholderKeys = new Map();
+    const placeholderElements = new Map();
+
+    placeholderAppearance.forEach((placeholderAppearances, placeholderName) => {
+      placeholderKeys.set(
+        placeholderName,
+        `__PLACEHOLDER__${placeholderName}__${uniqueId()}__`
+      );
+      placeholderElements.set(
+        placeholderName,
+        placeholderAppearances.map(() => document.createElement("span"))
+      );
+    });
+
+    const text = i18n(this.args.key, {
+      ...Object.fromEntries(placeholderKeys),
+      ...optionsArg,
+    });
+
+    // Bail early if there were no placeholders we need to handle.
+    if (placeholderAppearance.size === 0) {
+      return {
+        appearance: placeholderAppearance,
+        elements: placeholderElements,
+        keys: placeholderKeys,
+        rendered: [],
+        parts: [text],
+      };
+    }
```

Comment:
If it were me, I'd probably just skip this early return. The next piece of code will be a no-op when there are no placeholders, so it seems fine to follow the same path for simplicity 🤷‍♂️ 

(non-blocking comment - happy to keep this if you prefer)